connectors/sources/mysql.py

```diff
@@ -111,31 +120,76 @@ async def ping(self):
+        """
 
-                while True:
-                    rows = await cursor.fetchmany(size=size)
+        size = int(self.configuration.get("fetch_size", DEFAULT_FETCH_SIZE))
+        logger.debug(f"Streaming {database}.{table} {size} rows at a time")
 
-                    if len(rows) == 0:
+        # retry: Current retry counter
+        # yield_once: Yield(fields) once flag
+        retry = yield_once = 1
+
+        # rows_fetched: Rows fetched counter
+        # cursor_position: Mocked cursor position
+        rows_fetched = cursor_position = 0
+        while retry <= self.retry_count:
+            try:
+                async with self.connection_pool.acquire() as connection:
+                    async with connection.cursor(aiomysql.cursors.SSCursor) as cursor:
+                        await cursor.execute(query)
+
+                        # sending back column names only once
+                        if yield_once:
+                            yield [column[0] for column in cursor.description]
+                            yield_once = 0
+
+                        # setting cursor position where it was failed
+                        if cursor_position:
+                            await cursor.scroll(cursor_position, mode="absolute")
+                            logger.debug(
+                                f"Cursor moved to {cursor_position} position for {database}.{table}"
+                            )
+
+                        while True:
+                            rows = await cursor.fetchmany(size=size)
+                            rows_length = len(rows)
+
+                            # resetting cursor position & retry to 0 for next batch
+                            if cursor_position:
+                                cursor_position = retry = 0
+
+                            if not rows_length:
+                                break
+
+                            for row in rows:
+                                yield row
+
+                            rows_fetched += rows_length
+                            await asyncio.sleep(0)
                         break
-
-                    for row in rows:
-                        yield row
-
-                    count += len(rows)
-                    if count > 100:
-                        logger.info(f"Collected {count} rows in {database}.{table}")
-                        count = 0
-                    await asyncio.sleep(0)
+            except IndexError as exception:
+                logger.exception(
+                    f"None of rows fetched from {rows_fetched} rows in {database}.{table}. Exception: {exception}"
+                )
+                break
+            except Exception as exception:
+                logger.warn(
+                    f"Retry count: {retry} out of {self.retry_count} for rows in {database}.{table}. Exception: {exception}"
+                )
+                if retry == self.retry_count:
```

Comment:
adding the raise here and doing breaks make the code very hard to read. I also think that you might get connection issues when you call MySQL outside of that specific loop in other parts of the class.

I would like to suggest refactoring this a bit by creating a dedicated decorator function with the retry mechanism

something like

```
def _with_retries(self, callable, *args, **kw):
    retries = 0
    while True:
        try:
            return callable(*args, **kw)
        except CatchSomeError:
            retries += 1
            if retries >= self.retry_count:
                raise
            sleep...
```

and then in your code:

```
await self._with_retries(self._stream_rows, database, table, query)

```

Reply:
Hey @tarekziade , We tried this approach, there are few things that we want to discuss.

- We won't be able to return the callable directly since, the callable will return the following:

1. **Generator**: the function stream_rows returns a generator. Also, we cannot return the generator directly as well since the value of the generator can be:
    - Results: In case of the successful response, the callable will return a generator with the results of the query
    - Exception: In case of a server exception, the callable will return a generator with the exception.
Hence, we'll have to iterate over the generator returned by the callable to verify if the generator contains an Exception, then retry the connection again.

2. **Direct Response(List/Tuple)**: The function execute_query directly returns the response in the form of List or Tuple. 

   
Hence we'll to tweak this function as below:

```
def _with_retries(self, callable, *args, **kw):
    retries = 0
    while True:
        try:
            if returntype(callable) == <generator>:
                for item in generator:
                    if exception:
                        retry
                    else return
            else:
                return callable(*args, **kw)
        except CatchSomeError:
            retries += 1
            if retries >= self.retry_count:
                raise
            sleep...
```

Please let us know your thoughts for the same

Reply:
I think the easiest way to do this would be to retry only at the query level, not the generator level. 

So every method from the `cursor` object -- could be transparent to the connector's code.

One thing you could do is create a new class from `aiomysql.cursors.SSCursor` and add the retry mechanism for `fetchmany`, `scroll` and `execute`

and then use it in `async with connection.cursor(MyCustomSSCursor) as cursor:`

Reply:
What I am not sure about is if you can actually retry a `cursor` call

Reply:
@tarekziade We tried several ways to manage the connection and cursor to handle retry. However, one issue with retrying inside the Cursor class is that on an exception, the current connection will be lost and if we try to create a new connection to retry, the reference b/w the cursor and connection will be lost. Hence, it doesn't seem reasonable to retry(re-create connection) inside a cursor class. We even tried assigning a new connection to the existing cursor inside the SSCursor class but the _reader fails initialization.

Since the connector is public and people might be using it, do you think it makes sense to get the basic retry functionality merged and then work on the future enhancement to make it generic?

Reply:
sure ok, that sounds fair. The current retry is only on document retrievals, though. we would still need to have a retry for the other calls made around that loop.  I propose that you create a new bug explaining the follow up work, with a list of places that need a retry